Sources/PostgresNIO/New/Data/Int+PSQLCodable.swift

```diff
@@ -0,0 +1,167 @@
+        byteBuffer.writeInteger(self, as: Int16.self)
+    }
+}
+
+extension Int32: PSQLCodable {
+    var psqlType: PSQLDataType {
+        .int4
+    }
+    
+    // decoding
+    static func decode(from buffer: inout ByteBuffer, type: PSQLDataType, context: PSQLDecodingContext) throws -> Self {
+        switch type {
+        case .int2:
+            guard buffer.readableBytes == 2, let value = buffer.readInteger(as: Int16.self) else {
+                throw PSQLCastingError.failure(targetType: Self.self, type: type, postgresData: buffer, context: context)
+            }
+            
+            return Int32(value)
+        case .int4:
+            guard buffer.readableBytes == 4, let value = buffer.readInteger(as: Int32.self) else {
+                throw PSQLCastingError.failure(targetType: Self.self, type: type, postgresData: buffer, context: context)
+            }
+            
+            return Int32(value)
+        default:
+            throw PSQLCastingError.failure(targetType: Self.self, type: type, postgresData: buffer, context: context)
+        }
+    }
+    
+    // encoding
+    func encode(into byteBuffer: inout ByteBuffer, context: PSQLEncodingContext) {
+        byteBuffer.writeInteger(self, as: Int32.self)
+    }
+}
+
+extension Int64: PSQLCodable {
+    var psqlType: PSQLDataType {
+        .int8
+    }
+    
+    // decoding
+    static func decode(from buffer: inout ByteBuffer, type: PSQLDataType, context: PSQLDecodingContext) throws -> Self {
+        switch type {
+        case .int2:
+            guard buffer.readableBytes == 2, let value = buffer.readInteger(as: Int16.self) else {
+                throw PSQLCastingError.failure(targetType: Self.self, type: type, postgresData: buffer, context: context)
+            }
+            
+            return Int64(value)
+        case .int4:
+            guard buffer.readableBytes == 4, let value = buffer.readInteger(as: Int32.self) else {
+                throw PSQLCastingError.failure(targetType: Self.self, type: type, postgresData: buffer, context: context)
+            }
+            
+            return Int64(value)
+        case .int8:
+            guard buffer.readableBytes == 8, let value = buffer.readInteger(as: Int64.self) else {
+                throw PSQLCastingError.failure(targetType: Self.self, type: type, postgresData: buffer, context: context)
+            }
+            
+            return value
+        default:
+            throw PSQLCastingError.failure(targetType: Self.self, type: type, postgresData: buffer, context: context)
+        }
+    }
+    
+    // encoding
+    func encode(into byteBuffer: inout ByteBuffer, context: PSQLEncodingContext) {
+        byteBuffer.writeInteger(self, as: Int64.self)
+    }
+}
+
+extension Int: PSQLCodable {
+    var psqlType: PSQLDataType {
+        #if (arch(i386) || arch(arm))
+        return .int4
+        #else
+        return .int8
+        #endif
```

Comment:
This check is insufficient.
```suggestion
        switch Int.bitWidth {
        case Int32.bitWidth: return .int4
        case Int64.bitWidth: return .int8
        default: fatalError("No support for `Int` being \(Int.bitWidth) bits")
        }
```
There's at least one other place in the code where this is checked this way.

Reply:
Would you mind explaining why this is insufficient? Why should we evaluate this at runtime and not at compile time...

Reply:
To put it simply, it's insufficient to cover the set of architectures Swift is expected to be capable of running on - nor is it guaranteed that `Int` is always 4 bytes on 32-bit platforms or 8 bytes on 64-bit platforms.

In addition, `Int.bitWidth` is a compile-time constant, `public`, and `@inlinable` (in effect if not necessarily explicitly in the stdlib interface, due to compiler magic); in release builds (and possibly even in debug builds) this will always be lifted to a constant return value by the optimizer.